stl/src/tzdb.cpp

```diff
@@ -1,15 +1,297 @@
+    //    _Info->_Err == _Icu_error --> runtime_error interacting with ICU
+    _STD unique_ptr<__std_tzdb_time_zones_info, decltype(&__std_tzdb_delete_time_zones)> _Info{
+        new (_STD nothrow) __std_tzdb_time_zones_info{}, &__std_tzdb_delete_time_zones};
+    if (_Info == nullptr) {
+        return nullptr;
+    }
+
+    if (_Acquire_icu_functions() < _Icu_api_level::__has_icu_addresses) {
+        _Info->_Err = __std_tzdb_error::_Win_error;
+        return _Info.release();
+    }
+
+    UErrorCode _Err{};
+    _STD unique_ptr<UEnumeration, decltype(&__icu_uenum_close)> _Enum{
+        __icu_ucal_openTimeZoneIDEnumeration(USystemTimeZoneType::UCAL_ZONE_TYPE_CANONICAL, nullptr, nullptr, &_Err),
+        &__icu_uenum_close};
+    if (U_FAILURE(_Err)) {
+        _Info->_Err = __std_tzdb_error::_Icu_error;
+        return _Info.release();
+    }
+
+    // uenum_count may be expensive but is required to pre allocated arrays.
+    int32_t _Num_time_zones = __icu_uenum_count(_Enum.get(), &_Err);
+    if (U_FAILURE(_Err)) {
+        _Info->_Err = __std_tzdb_error::_Icu_error;
+        return _Info.release();
+    }
+
+    _Info->_Num_time_zones = static_cast<size_t>(_Num_time_zones);
+    _Info->_Names          = new (_STD nothrow) const char*[_Info->_Num_time_zones];
+    if (_Info->_Names == nullptr) {
+        return nullptr;
+    }
+
+    // init to ensure __std_tzdb_delete_init_info cleanup is valid
+    _STD fill_n(_Info->_Names, (ptrdiff_t) _Info->_Num_time_zones, nullptr);
+
+    _Info->_Links = new (_STD nothrow) const char*[_Info->_Num_time_zones];
+    if (_Info->_Links == nullptr) {
+        return nullptr;
+    }
+
+    // init to ensure __std_tzdb_delete_init_info cleanup is valid
+    _STD fill_n(_Info->_Links, _Info->_Num_time_zones, nullptr);
+
+    for (size_t _Name_idx = 0; _Name_idx < _Info->_Num_time_zones; ++_Name_idx) {
+        int32_t _Elem_len{};
+        const auto* _Elem = __icu_uenum_unext(_Enum.get(), &_Elem_len, &_Err);
+        if (U_FAILURE(_Err) || _Elem == nullptr) {
+            _Info->_Err = __std_tzdb_error::_Icu_error;
+            return _Info.release();
+        }
+
+        _Info->_Names[_Name_idx] = _Allocate_wide_to_narrow(_Elem, _Elem_len, _Info->_Err);
+        if (_Info->_Names[_Name_idx] == nullptr) {
+            return _Info->_Err != __std_tzdb_error::_Success ? _Info.release() : nullptr;
+        }
+
+        // ensure time_zone is not a known time_zone_link
+        for (const auto& _Link : _Known_links) {
+            if (strcmp(_Info->_Names[_Name_idx], _Link._Name) == 0) {
+                _Info->_Links[_Name_idx] = _Link._Target; // no need to allocate a string
+            }
+        }
+    }
+
+    return _Info.release();
+}
+
+void __stdcall __std_tzdb_delete_time_zones(__std_tzdb_time_zones_info* _Info) noexcept {
+    if (_Info != nullptr) {
+        if (_Info->_Names != nullptr) {
+            for (size_t _Idx = 0; _Idx < _Info->_Num_time_zones; _Idx++) {

```

Comment:
We conventionally prefer preincrement by default, which makes any necessary postincrements stand out more.